### PR TITLE
SONARJAVA-3632 Rule S6104: Map "computeIfAbsent()" should not be used to add "null" values.

### PR DESCRIPTION
**Pull request by dorian-burihabwa-sonarsource**

### CODE REVIEW

File: java-checks/src/test/java/org/sonar/java/checks/NullReturnedOnComputeIfPresentOrAbsentCheckTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.jupiter.api.Test;
+import org.sonar.java.checks.verifier.JavaCheckVerifier;
+
+import static org.sonar.java.CheckTestUtils.testSourcesPath;
+
+class NullReturnedOnComputeIfPresentOrAbsentCheckTest {
+  @Test
+  void test() {
+    JavaCheckVerifier
+      .newVerifier()
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/NullReturnedOnComputeIfPresentOrAbsentCheckTest.java:31]: Same here idea as the comment in the main code, you don't need to separate the method call from the type.

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/java/S6104_java.html [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+<pre>
+if (!map.containsKey(key)) {
+    map.put(key, null);
+}
+if (map.containsKey(key)) {
+    map.put(key, null);
+}
+</pre>
+<h2>See also</h2>
+<p> * {rule:squid:S3824} - "Map.get" and value test should be replaced with single method call</p>
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/java/S6104_java.html:24]: This is a bug in `ruleapi`, this should be an HTML list. 
This is due to the fact that a space is sometimes inserted automatically before the `*` in the [rspec](https://jira.sonarsource.com/browse/RSPEC-6104). I updated the description to fix it, you can re-generate the description.

**Review by quentin-jaquier-sonarsource — CHANGES_REQUESTED** [review]
In addition, I think you could have squashed some of the commits of your branch. It makes sense to separate some of them to keep a history of what you tried, help the reviewer or intent to keep all the commit separated in master.

**Comment by dorian-burihabwa-sonarsource** [comment]
If the code changes are satisfactory enough, I can squash and rebase on master

File: java-checks/src/main/java/org/sonar/java/checks/NullReturnedOnComputeIfPresentOrAbsentCheck.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+public class NullReturnedOnComputeIfPresentOrAbsentCheck extends AbstractMethodDetection {
+  private static final String PRIMARY_MESSAGE = "Use \"Map.containsKey(key)\" followed by \"Map.put(key, null)\" to add null values.";
+  private static final String SECONDARY_MESSAGE = "null literal in the arguments";
+  private static final MethodMatchers COMPUTE_IF = MethodMatchers.create()
+    .ofTypes("java.util.Map")
+    .names("computeIfPresent", "computeIfAbsent")
+    .addParametersMatcher(MethodMatchers.ANY, MethodMatchers.ANY)
+    .build();
+
+  @Override
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/NullReturnedOnComputeIfPresentOrAbsentCheck.java:45]: This annotation should go after the javadoc.

**Review by quentin-jaquier-sonarsource — COMMENTED** [review]
A last small formatting comment, otherwise, LGTM!

**Review by quentin-jaquier-sonarsource — APPROVED** [review]
LGTM, well done!

**Pull request merged**